README.md
# horizontal_list

[![pub](https://img.shields.io/pub/v/horizontal_list)](https://pub.dev/packages/horizontal_list)
[![license](https://img.shields.io/badge/license-mit-green.svg)](https://github.com/Dansp/horizontal_list/blob/main/LICENSE)
<!-- [![build](https://github.com/Dansp/horizontal_list/actions/workflows/build.yml/badge.svg)](https://github.com/Dansp/horizontal_list/actions/) -->
[![codecov](https://codecov.io/gh/Dansp/horizontal_list/branch/main/graph/badge.svg?token=jYfO55O22s)](https://codecov.io/gh/Dansp/horizontal_list)
[![CodeFactor Grade](https://img.shields.io/codefactor/grade/github/Dansp/horizontal_list)](https://www.codefactor.io/repository/github/Dansp/horizontal_list)
[![Hits-of-Code](https://hitsofcode.com/github/Dansp/horizontal_list?branch=main)](https://hitsofcode.com/github/Dansp/horizontal_list/view?branch=main)

A horizontal list widget with buttons next and previous. You can customize your card for example and insert it on HorizontalListView widget.


<div>
  WEB/DESKTOP
  <br>
  <img src="https://raw.githubusercontent.com/Dansp/horizontal_list/main/media/web.gif" height="400">
  <br>
  APP
  <br>
  <img src="https://raw.githubusercontent.com/Dansp/horizontal_list/main/media/mobile.gif" height="400">
</div>

<br>

## Using the library

Add the repo to your Flutter `pubspec.yaml` file.

```
dependencies:
  horizontal_list: <<version>>
```

Then run...
```
flutter packages get
```


## Example

```dart
HorizontalListView(
    width: double.maxFinite, //Width of widget
    height: 200, //Height of widget
    list: [Text('Text 1'), Text('Text 2')], //List of widget
    iconNext: Icon(Icons.arrow_forward_ios), // Icon for button next
    iconPrevious: Icon(Icons.arrow_back_ios), // Icon for button previous
    curveAnimation: Curves.bounceIn, //Curve for animation
    durationAnimation: Duration(milliseconds: 300), //Duration of animation
    enableManualScroll: true, //Enable manual scroll
    onNextPressed: () { //On button next pressed
      print('On next pressed');
    },
    onPreviousPressed: () { //On button next pressed
      print('On previous pressed');
    },
),
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Dart Package Versioning](https://dart.dev/tools/pub/versioning).


## [0.0.7] - 2023-02-09

### Added

- onChanged listener with index added;
- Make icons always visible if it has.

## [0.0.6] - 2023-02-09

### Added

- Changed `scrollSize` to double.

## [0.0.5] - 2022-02-23

### Added

- Log removed;
- readme adjusted.

## [0.0.4] - 2022-02-17

### Added

- Readme badges added;
- enableManualScroll added:
    - If true, enable manual scroll, if false disable manual scroll and can only be scrolled by iconNext and iconPrevious.

## [0.0.3] - 2022-02-02

### Added

- Bug fix on scroll;

## [0.0.2+1] - 2022-02-02

### Added

- description changed;

## [0.0.2] - 2022-02-02

### Added

- example project;
- insert size height and width for Widget

## [0.0.1] - 2022-02-01

### Added

- Create the lib horizontal_list with widget HorizontalListView

test/horizontal_list_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:horizontal_list/horizontal_list.dart';

void main() {
  Widget makeTestable(Widget widget) => MaterialApp(home: widget);

  testWidgets('test my widget', (WidgetTester tester) async {
    await tester.pumpWidget(makeTestable(const HorizontalListView(
      height: 200,
      width: double.maxFinite,
      iconNext: Icon(Icons.arrow_forward_ios),
      iconPrevious: Icon(Icons.arrow_back_ios),
      list: [Text('1'), Text('2')],
    )));
    await tester.pumpAndSettle(const Duration(seconds: 3));
    final l1Finder = find.text('1');
    final l2Finder = find.text('2');
    final l3Finder = find.text('3');

    expect(l1Finder, findsOneWidget);
    expect(l2Finder, findsOneWidget);
    expect(l3Finder, findsNothing);
  });
}


lib/horizontal_list.dart
library horizontal_list;

import 'dart:async';

import 'package:flutter/material.dart';

class HorizontalListView extends StatefulWidget {
  ///Height of widget
  final double height;

  ///Width of widget
  final double width;

  ///List of widget to populate horizontal scroll.
  final List<Widget> list;

  ///Icon to show in next button
  final Icon? iconPrevious;

  ///Icon to show in next button
  final Icon? iconNext;

  ///On click listener when index changed;
  ///NOTE: Only will work with `enableManualScroll` false.
  final Function(int index)? onChanged;

  ///On click listener when next button was clicked.
  final VoidCallback? onNextPressed;

  ///On click listener when previous button was clicked.
  final VoidCallback? onPreviousPressed;

  ///Duration of animation when click on button next or previous. Default is 500 milliseconds.
  ///
  /// Duration(milliseconds: 500);
  final Duration durationAnimation;

  ///Curve to animation when click on button next or previous. Default is [Curves.ease]
  final Curve curveAnimation;

  ///Item width.
  ///
  ///The width of item in list. Default is 300.
  final double itemWidth;

  ///Start from end of list
  ///
  ///Default is false;
  final bool isStartedFromEnd;

  ///If true, enable manual scroll, if false disable manual scroll and can only be scrolled by [iconNext] and [iconPrevious].
  ///
  /// Default is true.
  final bool enableManualScroll;

  const HorizontalListView({
    required this.height,
    required this.width,
    required this.list,
    this.itemWidth = 300,
    this.iconPrevious,
    this.iconNext,
    this.onChanged,
    this.onNextPressed,
    this.onPreviousPressed,
    this.durationAnimation = const Duration(milliseconds: 500),
    this.curveAnimation = Curves.ease,
    this.enableManualScroll = true,
    this.isStartedFromEnd = false,
    Key? key,
  }) : super(key: key);

  @override
  State<HorizontalListView> createState() => _HorizontalScrollState();
}

class _HorizontalScrollState extends State<HorizontalListView> {
  final ScrollController _controller = ScrollController();
  int _index = 0;
  var _reachEnd = true;
  var _startScroll = false;

  _listener() {
    if (_controller.hasClients) {
      final maxScroll = _controller.position.maxScrollExtent;
      final minScroll = _controller.position.minScrollExtent;
      setState(() {
        if (_controller.offset <= minScroll && maxScroll > 0) {
          _reachEnd = false;
        }
        _reachEnd = _controller.offset >= maxScroll;
        _startScroll = _controller.offset > minScroll;
      });
    }
  }

  @override
  void initState() {
    _controller.addListener(_listener);
    super.initState();
    if (!_controller.hasClients) {
      _waitWhile(() => !_controller.hasClients).then((value) {
        if (widget.isStartedFromEnd) {
          _index = widget.list.length - 1;
          _controller.jumpTo(_index * (widget.itemWidth - 100));

          if (!widget.enableManualScroll) {
            widget.onChanged?.call(_index);
          }
        }
        _listener();
      });
    }
  }

  ///Wait to wait a bool variable to change
  Future _waitWhile(bool Function() test,
      [Duration pollInterval = Duration.zero]) {
    var completer = Completer();
    check() {
      if (!test()) {
        completer.complete();
      } else {
        Timer(pollInterval, check);
      }
    }

    check();
    return completer.future;
  }

  @override
  void dispose() {
    _controller.removeListener(_listener);
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: widget.width,
      height: widget.height,
      child: Stack(children: [
        Positioned(
          top: 0.0,
          bottom: 0,
          left: widget.iconPrevious != null ? 50.0 : 0,
          right: widget.iconNext != null ? 50.0 : 0,
          child: ListView.builder(
            physics: widget.enableManualScroll
                ? const ClampingScrollPhysics()
                : const NeverScrollableScrollPhysics(),
            scrollDirection: Axis.horizontal,
            controller: _controller,
            shrinkWrap: true,
            itemBuilder: (context, index) {
              return SizedBox(
                  width: widget.itemWidth - 100, child: widget.list[index]);
            },
            itemCount: widget.list.length,
          ),
        ),
        widget.iconPrevious != null
            ? Align(
                alignment: Alignment.centerLeft,
                child: IconButton(
                  icon: widget.iconPrevious!,
                  onPressed: _startScroll
                      ? () async {
                          _index -= 1;
                          if (_index < 0) {
                            _index = 0;
                          }

                          await _controller.animateTo(
                              _index * (widget.itemWidth - 100),
                              duration: widget.durationAnimation,
                              curve: widget.curveAnimation);
                          if (!widget.enableManualScroll) {
                            widget.onChanged?.call(_index);
                          }
                          widget.onPreviousPressed?.call();
                        }
                      : null,
                ))
            : Container(),
        widget.iconNext != null
            ? Align(
                alignment: Alignment.centerRight,
                child: IconButton(
                  icon: widget.iconNext!,
                  onPressed: _reachEnd
                      ? null
                      : () async {
                          _index += 1;
                          if (_index > widget.list.length - 1) {
                            _index = widget.list.length - 1;
                          }
                          await _controller.animateTo(
                              _index * (widget.itemWidth - 100),
                              duration: widget.durationAnimation,
                              curve: widget.curveAnimation);

                          if (!widget.enableManualScroll) {
                            widget.onChanged?.call(_index);
                          }
                          widget.onNextPressed?.call();
                        },
                ))
            : Container(),
      ]),
    );
  }
}


example/lib/main.dart
import 'dart:developer';

import 'package:flutter/material.dart';
import 'package:horizontal_list/horizontal_list.dart';
import 'dart:math' as math;

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Horizontal list Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Horizontal list Example'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Widget> _itemsComponent() {
    List<Widget> myList = [];
    for (var i = 0; i < 7; i++) {
      myList.add(Container(
          width: 300,
          color: Color((math.Random().nextDouble() * 0xFFFFFF).toInt())
              .withOpacity(1.0),
          child: Center(
              child: Text('Card ${i + 1}',
                  style: const TextStyle(fontWeight: FontWeight.bold)))));
    }
    return myList;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: Text(widget.title),
        ),
        body: Column(
          children: [
            const SizedBox(height: 150),
            HorizontalListView(
              width: double.maxFinite,
              height: 200,
              list: _itemsComponent(),
              iconPrevious: const Icon(Icons.arrow_back_ios),
              iconNext: const Icon(Icons.arrow_forward_ios),
              isStartedFromEnd: false,
              itemWidth: 300,
              onChanged: (index) {
                print(index);
              },
              onPreviousPressed: () {
                //DO WHAT YOU WANT ON PREVIOUS PRESSED
                log('onPreviousPressed');
              },
              onNextPressed: () {
                //DO WHAT YOU WANT ON NEXT PRESSED
                log('onNextPressed');
              },
            ),
          ],
        ));
  }
}
